### ISSUE TITLE
How to use FactoryAggregate in a container

### ISSUE THREAD

**Issue by githubkusi** [ets-labs/python-dependency-injector#239]
Can I use FactoryAggregate in a DeclarativeContainer, where I choose the proper factory via config? The following code does not work

```python
from dependency_injector import containers, providers


class A(object):
    name = "A"


class B:
    name = "B"


class C:
    def __init__(self, a_or_b):
        print(a_or_b.name)


class IoC(containers.DeclarativeContainer):
    config = providers.Configuration('config')
    aggregate_factory = providers.FactoryAggregate(a=providers.Factory(A), b=providers.Factory(B))
    # c_factory = providers.Factory(C, aggregate_factory('a')) # this works, but ignores the config
    c_factory = providers.Factory(C, aggregate_factory(config.a_or_b)) # this fails


def main():
    container = IoC(config={'a_or_b': 'a'})
    c = container.c_factory()


main()
```

**Comment by rmk135** [comment]
Hi @githubkusi ,

This will not work, unfortunately. However, this is an interesting feature that would be nice to have in a framework.

What we can do is looking for a short-term spike how to get it work with a further building into the next version of DI.

I'll send a sample code that would help to make it work shortly.

Thanks,
Roman

**Comment by rmk135** [comment]
Here we go, that works:

```python
from dependency_injector import containers, providers


class FactoryAggregateSelector(providers.Provider):

    def __init__(self, factory_aggregate, factory_name_provider):
        self._factory_aggregate = factory_aggregate
        self._factory_name_provider = factory_name_provider
        super(providers.Provider, self).__init__()

    def __deepcopy__(self, memo):
        """Create and return full copy of provider."""
        copied = memo.get(id(self))
        if copied is not None:
            return copied

        copied = self.__class__(
            providers.deepcopy(self._factory_aggregate, memo),
            providers.deepcopy(self._factory_name_provider, memo),
        )

        self._copy_overridings(copied, memo)

        return copied

    def _provide(self, args, kwargs):
        return self._factory_aggregate(self._factory_name_provider(), *args, **kwargs)


class A(object):
    name = "A"


class B:
    name = "B"


class C:
    def __init__(self, a_or_b):
        print(a_or_b.name)


class IoC(containers.DeclarativeContainer):
    config = providers.Configuration('config')
    aggregate_factory = providers.FactoryAggregate(a=providers.Factory(A), b=providers.Factory(B))
    c_factory = providers.Factory(C, FactoryAggregateSelector(aggregate_factory, config.a_or_b)) # this will work


def main():
    container = IoC(config={'a_or_b': 'a'})
    c = container.c_factory()


main()
```

**Comment by rmk135** [comment]
That would be a useful feature, should definitely be builtin. Thanks for sharing the issue @githubkusi 

Best Regards,
Roman

**Comment by githubkusi** [comment]
this works, thanks for the quick solution! Hopefully this goes into the next version

**Comment by githubkusi** [comment]
unfortunately, this doesn't work when "a" is a singleton

**Comment by rmk135** [comment]
That happens because `FactoryAggregate` provider can only accept `Factory` providers. This is a really dirty hack, but that's how this could work if you need singletons:

```
class IoC(containers.DeclarativeContainer):
    config = providers.Configuration('config')
    aggregate_factory = providers.FactoryAggregate(
        a=providers.Factory(providers.Singleton(A)),
        b=providers.Factory(providers.Singleton(B)),
    )
    c_factory = providers.Factory(C, FactoryAggregateSelector(aggregate_factory, config.a_or_b))
```

**Comment by rmk135** [comment]
@githubkusi 

The time has come.

I've added new `Selector` provider. It covers the case from this issue. It also supports `Singleton` or any other provider.

Code from the issue now can look like:

```python
from dependency_injector import containers, providers


class IoC(containers.DeclarativeContainer):
    config = providers.Configuration('config')
    factory_selector = providers.Selector(config.a_or_b, a=providers.Singleton(A), b=providers.Singleton(B))
    c_factory = providers.Factory(C, factory_selector)


def main():
    container = IoC(config={'a_or_b': 'a'})
    c = container.c_factory()


if __name__ == '__main__':
    main()
```

- `Selector` provider is available since version `3.19`. It's already on PyPI.
- Here is a doc page for - [`Selector` provider](http://python-dependency-injector.ets-labs.org/providers/selector.html).


Thank you for bringing up the issue.

Roman
